ssc/relaxation.py

```diff
@@ -0,0 +1,125 @@
+import numpy
+from firedrake.petsc import PETSc
+
+__all__ = ['PlaneSmoother', 'OrderedVanka']
+
+class PlaneSmoother(object):
+    @staticmethod
+    def coords(dm, p):
+        coordsSection = dm.getCoordinateSection()
+        coordsDM = dm.getCoordinateDM()
+        dim = coordsDM.getDimension()
+        coordsVec = dm.getCoordinatesLocal()
+        return dm.getVecClosure(coordsSection, coordsVec, p).reshape(-1, dim).mean(axis=0)
+
+    def sort_entities(self, dm, axis, dir, ndiv):
+        # compute
+        # [(pStart, (x, y, z)), (pEnd, (x, y, z))]
+        entities = [(p, self.coords(dm, p)) for p in range(*dm.getChart())]
+
+        minx = min(entities, key=lambda z: z[1][axis])[1][axis]
+        maxx = max(entities, key=lambda z: z[1][axis])[1][axis]
+        def keyfunc(z):
+            coords = tuple(z[1])
+            return (coords[axis], ) + tuple(coords[:axis] + coords[axis+1:])
+        s = sorted(entities, key=keyfunc, reverse=(dir==-1))
+
+        divisions = numpy.linspace(minx, maxx, ndiv+1)
+        (entities, coords) = zip(*s)
+        coords = [c[axis] for c in coords]
+        indices = numpy.searchsorted(coords[::dir], divisions)
+
+        out = []
+        for k in range(ndiv):
+            out.append(entities[indices[k]:indices[k+1]])
+        out.append(entities[indices[-1]:])
+
+        return out
+
+    def __call__(self, pc):
+        dm = pc.getDM()
+        prefix = pc.getOptionsPrefix()
+        sentinel = object()
+        sweeps = PETSc.Options(prefix).getString("pc_patch_construction_ps_sweeps", default=sentinel)
```

Comment:
TODO, how to add the relevant `-help` string.

Reply:
It looks like it's not supported in petsc4py.